language-plutus-core/src/Language/PlutusCore/Constant/Dynamic/Instances.hs

```diff
@@ -47,6 +54,108 @@ updated state that the first 'eval' finished with. This may cause all kinds of w
 for example, an error message saying that there is a free variable and evaluation cannot proceed.
 -}
 
+-- | A wrapper around 'Term' parameterized by a phantom type variable representing the Haskell
+-- counterpart of the type of that term.
+newtype TypedTerm a = TypedTerm
+    { unTypedTerm :: Term TyName Name ()
+    }
+
+{- Note [Dynamic builtins as constructors]
+We can use dynamic builtins as improvised constructors of data types.
+Consider the 'Sealed' example, we have:
+
+1. the hardcoded @sealed@ PLC type of kind @* -> *@
+2. the @seal@ dynamic builtin of type @all a. a -> sealed a@
+3. the @unseal@ dynamic builtin of type @all a. sealed a -> a@
+
+The Haskell meanings of these things are:
+
+1. 'Sealed' (data type)
+2. 'Sealed' (constructor)
+3. 'unSealed' (field selector)
+
+I.e.
+
+    seal {Integer} 1
+
+is a value of type @sealed integer@, i.e. @seal@ plays the role of a constructor here,
+even though it's really a dynamic builtin.
+
+We immediately run into problems, though. As a dynamic builtin, @seal@ has a meaning and thus
+computes. So @seal@ becomes @Sealed@ on the Haskell side, but we also need to go in the other
+direction, meaning that @Sealed@ becomes @seal@ back. I.e. the
+
+    seal {Integer} 1
+
+application evaluates to itself, which means that the previous version of the CEK machine was
+looping while trying to evaluate that application over and over again. The hack that we added
+to handle this is trivial: if an application evaluates to itself, do not attempt to evaluate it
+again.
+
+The other difficulty that we have is related to unlifting of values that look like that.
+Normally, during evaluation we get one of these:
+
+- an actual constructor (via 'IWrap')
+- a ground value (via 'Constant')
+- a function (via 'Lam')
```

Comment:
Making some builtin applications "stuck" also breaks all of our stuff around values. Builtin applications aren't values, so things of type `sealed a` are never going to be values, which is very weird.

Reply:
There is an alternative way of doing this: define `sealed a = () -> sealedStuck a` with `sealedStuck` being a builtin type. Then values of type `sealed a` are not only proper values, but we also don't need to add the hack to the CEK machine, because evaluation of `\x -> b` results in `\x -> b` and halts.

Reply:
What happens if you ever try to evaluate a `sealedStuck`? Then it loops and we just hope that never happens?

Reply:
We can probably throw an error in this case. While this is a good point, I think it's not a big deal, because if the user is not attempting to cheat and just uses our primitives (or a fair compiler), then they won't run into this issue. Obviously, there are plenty of places where a compiler can screw up.